tests/attr/test_lime.py

```diff
@@ -0,0 +1,445 @@
+from captum.attr._core.lime import (
+    Lime,
+    LimeBase,
+    default_similarity_kernel,
+    lasso_interpretable_model_trainer,
+)
+from captum.attr._utils.common import (
+    _construct_default_feature_mask,
+    _format_input,
+    _format_input_baseline,
+)
+
+from ..helpers.basic import (
+    BaseTest,
+    assertTensorAlmostEqual,
+    assertTensorTuplesAlmostEqual,
+)
+from ..helpers.basic_models import (
+    BasicModel_MultiLayer,
+    BasicModel_MultiLayer_MultiInput,
+)
+
+
+def alt_sampling_func(
+    original_inp: TensorOrTupleOfTensorsGeneric, **kwargs
+) -> TensorOrTupleOfTensorsGeneric:
+    if isinstance(original_inp, Tensor):
+        device = original_inp.device
+    else:
+        device = original_inp[0].device
+
+    feature_mask = kwargs["feature_mask"]
+
+    probs = torch.ones(1, kwargs["num_interp_features"]) * 0.5
+    curr_sample = torch.bernoulli(probs).to(device=device)
+
+    binary_mask: TensorOrTupleOfTensorsGeneric
+    if isinstance(original_inp, Tensor):
+        binary_mask = curr_sample[0][feature_mask]
+        return binary_mask * original_inp + (1 - binary_mask) * kwargs["baselines"]
+    else:
+        binary_mask = tuple(
+            curr_sample[0][feature_mask[j]] for j in range(len(feature_mask))
+        )
+        return tuple(
+            binary_mask[j] * original_inp[j]
+            + (1 - binary_mask[j]) * kwargs["baselines"][j]
+            for j in range(len(feature_mask))
+        )
+
+
+def alt_to_interp_rep(
+    curr_sample: TensorOrTupleOfTensorsGeneric,
+    original_input: TensorOrTupleOfTensorsGeneric,
+    **kwargs: Any
+) -> Tensor:
+    binary_vector = torch.zeros(1, kwargs["num_interp_features"])
+    feature_mask = kwargs["feature_mask"]
+    for i in range(kwargs["num_interp_features"]):
+        curr_total = 1
+        if isinstance(curr_sample, Tensor):
+            if (
+                torch.sum(
+                    torch.abs(
+                        (feature_mask == i).float() * (curr_sample - original_input)
+                    )
+                )
+                > 0.001
+            ):
+                curr_total = 0
+        else:
+            sum_diff = sum(
+                torch.sum(torch.abs((mask == i).float() * (sample - inp)))
+                for inp, sample, mask in zip(original_input, curr_sample, feature_mask)
+            )
+            if sum_diff > 0.001:
+                curr_total = 0
+        binary_vector[0][i] = curr_total
+    return binary_vector
+
+
+class Test(BaseTest):
+    def test_simple_lime(self) -> None:
+        net = BasicModel_MultiLayer()
+        inp = torch.tensor([[20.0, 50.0, 30.0]], requires_grad=True)
+        self._lime_test_assert(
+            net,
+            inp,
+            [73.3716, 193.3349, 113.3349],
+            perturbations_per_eval=(1, 2, 3),
+            n_samples=500,
+            expected_coefs_only=[73.3716, 193.3349, 113.3349],
```

Comment:
Just commenting so I understand the tests better - is dependent on n_samples + seed + order that these tests execute as the random generator is not patched (due to the samples generated)?

Reply:
Good question! For most of these, I computed the "exact" expected output manually. In KernelSHAP, these are nicely exactly equal to true Shapley Values, which are straightforward to compute. For LIME, I manually constructed the expected distribution (containing all possible samples in the distribution weighted appropriately) and trained a Lasso regression model on that to obtain the expected results. The exact result does depend on n_samples, seed and order, but with sufficient samples, it should converge to within epsilon of the exact expected output. I've generally set the delta threshold and n_samples large enough to avoid any issue with random seed, but it's possible that if these parameters are too low for any case, the variance could be an issue.